### PR TITLE
[Widget enhancement] 17 - Emit toast for posting orders

### PR DESCRIPTION
**Pull request by anxolin**
# Summary

Part of the series https://github.com/cowprotocol/cowswap/pull/3812

This PR adds the first toast message: `Posting orders`. I had to make quite a big change because I needed the middleware to be stateless, and some relevant information was missing from the redux actions. 

## Details
To make the toast messages simpler for the clients, it will send a text message (as a `string`) so it can be passed to their message/toast library. 

This introduces a small problem, because we already have our toast messages, and we format using react components, if we want a string message version of the same, we will need to duplicate some logic. 

This PR tries to implement this first toast message, and for that it will need to:
- [x] Refactor the code to format token symbols/names (we don't use the `TokenSymbol` component, but it will use a similar formatting logic). Where possible, this PR refactors both codes
- [x] Refactor the code to format token amounts with currency (we don't use the `TokenAmount` component, but it will use a similar formatting logic). Where possible, this PR refactors both codes
- [x] Implement `getPendingOrderNotificationToast` to get the equivalent string version for the toast of `PendingOrderNotification` component
- [x] Implement `formatTokenAmountWithSymbol` to get the equivalent string version for the toast of `Token` component
- [x] Make sure the redux action payloads are self contained, so they don't rely on hooks or any other info later. This is good for some reasons:
    - Become state-less, for example, if we get a message for an order created by a safe (`isSafeWallet = true` ), and somehow we change the wallet to one that is not a safe (`isSafeWallet = true`), if we mix data from the payload from data from hooks then we can get into trouble
    - This was needed also, for implementing the message in a pure function that wouldn't depend on hooks

Additionally, this PR deletes some dead code and testing log messages

image

## Cosmos
Because now `PendingOrderNotification` is a dumb container and it needs to receive the `isSafeWallet` flag, I added the safe variations of the fixtures.

EOA:
image

Safe version
image

## Unit tests
Additionally adds some unit tests

image

## How to tests
1. Open the configurator: https://widget-configurator-git-widget-enhancement-event-99b848-cowswap.vercel.app
     - EDIT: You will need to test locally the configurator... see problems with the configurator link) 
3. Connect in Sepolia (or any network)
4. Show dev tools, and filter console events by `[configurator:`
5. Place an order
6. Check the console messages (should show something like the screenshot above)

## Issue with the links of the PREVIEW
I saw something is not working with the preview links in the configurator.

Is not part of this PR, so please test locally.

I suspect it has to do with the slash of the branch name.

Basically the good URL don't matches the one the configurator creates:
* THE ONE THIS PR CREATES: https://swap-dev-git-widget-enhancement-event-99b848-cowswap.vercel.app
* GOOD: https://swap-dev-git-widget-enhancement-events-and-params-17-cowswap.vercel.app

I checked and it has to do with the fact we assume the URL of the widget and cowswap generated by Vercel follows the same patter, but for some reason, in this PR doesnt
-  see https://github.com/cowprotocol/cowswap/blob/eff8657b05adc54b1d13eb58c08578cec1ebc45c/libs/widget-lib/src/consts.ts#L8

### CODE REVIEW

File: libs/widget-lib/src/IframeCowEventEmitter.ts [hunk 1]
```diff
@@ -33,10 +33,6 @@ export class IframeCowEventEmitter {
     if (event.data.key !== COW_SWAP_WIDGET_EVENT_KEY || event.data.method !== 'event') {
       return
     }
-    console.debug(
-      `[TODO:remove] Received message client side - Forward to eventEmitter ${event.data.eventName}`,
-      event.data.payload
-    )
```
Comment (anxolin) [inline libs/widget-lib/src/IframeCowEventEmitter.ts:39]: unrelated, just kills some test debug messages

File: apps/cowswap-frontend/src/common/containers/PendingOrderNotification/index.tsx [hunk 2]
```diff
@@ -28,26 +28,56 @@ const ORDER_TYPE_TITLES: Record<UiOrderType, string> = {
...
+    messageType: ToastMessageType.SWAP_POSTED_API,
+    message: toAddress && receiver ? `${baseMessage}. Receiver: ${toAddress}` : baseMessage,
+    data: {
+      orderUid: orderId,
+    },
+  }
+}
+
+export function PendingOrderNotification(props: PendingOrderNotificationProps) {
+  const { chainId, account, isSafeWallet, orderId, kind, orderType, inputAmount, outputAmount, receiver } = props
```
Comment (anxolin) [inline apps/cowswap-frontend/src/common/containers/PendingOrderNotification/index.tsx:80]: important, I had to make this component dumber. It was not good anyway that it was mustly dumb, but had some smart logic

Comment (anxolin) [reply apps/cowswap-frontend/src/common/containers/PendingOrderNotification/index.tsx:80]: The reason is, I will need the `getPendingOrderNotificationToast` to be self contained pure function and to have the same params as the component. So I can't rely on hooks

File: apps/cowswap-frontend/src/legacy/state/orders/actions.ts [hunk 3]
```diff
@@ -136,6 +136,7 @@ export interface AddPendingOrderParams {
   id: UID
   chainId: ChainId
   order: SerializedOrder
+  isSafeWallet: boolean
```
Comment (anxolin) [inline apps/cowswap-frontend/src/legacy/state/orders/actions.ts:139]: `isSafeWallet` is relevant later to give a full context on the order. This way, the middleware have everything they need to show the toast messages (or notify about the messages through the event emitter) without relying on hooks or any other dark magic. The message will be self contained

File: apps/cowswap-frontend/src/legacy/state/orders/middleware/popupMiddleware.ts [hunk 4]
```diff
@@ -31,6 +31,11 @@ export const popupMiddleware: Middleware<Record<string, unknown>, AppState> = (s
 
   if (isPendingOrderAction(action)) {
     pendingOrderPopup(store, action.payload)
+
+    // cowEventEmitter.emit(CowEvents.ON_TOAST_MESSAGE, {
+    //   messageType: ToastMessageType.SWAP_POSTED_API,
+    //   message: 'Swap Posted',
+    // })
```
Comment (anxolin) [inline apps/cowswap-frontend/src/legacy/state/orders/middleware/popupMiddleware.ts:38]: just kills dead code

File: apps/cowswap-frontend/src/legacy/state/orders/middleware/showPendingOrderNotification.tsx [hunk 5]
```diff
@@ -18,27 +25,48 @@ interface PendingOrderNotificationParams {
...
-          outputAmount={outputAmount}
-        />
-      ),
+      content,
     })
+
+    const toastMessage = getPendingOrderNotificationToast(pendingOrderNotificationMessage)
+    if (toastMessage) {
+      cowEventEmitter.emit(CowEvents.ON_TOAST_MESSAGE, toastMessage)
+    }
```
Comment (anxolin) [inline apps/cowswap-frontend/src/legacy/state/orders/middleware/showPendingOrderNotification.tsx:70]: Main addition to this PR. We just send the toast notification here

File: apps/cowswap-frontend/src/modules/injectedWidget/updaters/CowEventsUpdater.ts [hunk 6]
```diff
@@ -15,14 +15,16 @@ export function CowEventsUpdater() {
...
   // Setup listeners only once
   useEffect(() => {
-    // cowEventEmitter.on(CowEvents)
+    // Create all listeners
     const allHandlers: CowEventListeners = ALL_EVENTS.map((event) => {
       return {
         event,
-        handler: (payload: any) => forwardEventToIframe(event, payload),
+        handler: (payload: any) => {
+          console.debug('[CowEventsUpdater]: Forward event to iFrame', event, payload)
```
Comment (anxolin) [inline apps/cowswap-frontend/src/modules/injectedWidget/updaters/CowEventsUpdater.ts]: just some convenient debug message

File: apps/widget-configurator/src/app/configurator/index.tsx [hunk 7]
```diff
@@ -52,41 +52,45 @@ const COW_LISTENERS: CowEventListeners = [
...
-      }
+      console.info('[TODO:remove][configurator:toast:simple] 🍞 Message: ', event.message)
+    },
+  },
 
-      // ...or you can do handle them your way using the data:
+  {
+    event: CowEvents.ON_TOAST_MESSAGE,
+    handler: (event) => {
+      // You cn implement a more complex way to handle toast messages
```
Comment (anxolin) [inline apps/widget-configurator/src/app/configurator/index.tsx:63]: notice I just provide for illustration porpouses 2 ways to implement toast messages.

One is dumb (just show a message), the other is by handling the events

File: libs/ui/src/pure/TokenAmount/index.tsx [hunk 8]
```diff
@@ -23,7 +23,7 @@ export const SymbolElement = styled.span<{ opacitySymbol?: boolean }>`
 export interface TokenAmountProps {
   amount: Nullish<FractionLike>
   defaultValue?: string
-  tokenSymbol?: TokenSymbolProps['token']
+  tokenSymbol?: TokenNameAndSymbol
```
Comment (anxolin) [inline libs/ui/src/pure/TokenAmount/index.tsx:26]: I just gave a name to this type

File: apps/cowswap-frontend/src/common/pure/PendingOrderNotification/index.tsx [hunk 9]
```diff
@@ -28,26 +28,56 @@ const ORDER_TYPE_TITLES: Record<UiOrderType, string> = {
...
+  return {
+    messageType: ToastMessageType.SWAP_POSTED_API,
+    message: toAddress && receiver ? `${baseMessage}. Receiver: ${toAddress}` : baseMessage,
+    data: {
+      orderUid: orderId,
+    },
+  }
+}
+
+export function PendingOrderNotification(props: PendingOrderNotificationProps) {
```
Comment (anxolin) [inline apps/cowswap-frontend/src/common/pure/PendingOrderNotification/index.tsx:79]: `PendingOrderNotification` receives now all context it needs as props. As now is pure, I also moved it with the pure components

File: apps/cowswap-frontend/src/common/pure/PendingOrderNotification/PendingOrderNotification.test.ts [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+    chainId: SupportedChainId.MAINNET,
+    isSafeWallet: false,
+    kind: OrderKind.SELL,
+    orderId: '0x123',
+    orderType: UiOrderType.LIMIT,
+    ...overrides,
+  }
+}
+
+describe('Swap Orders', () => {
```
Comment (shoom3301) [inline apps/cowswap-frontend/src/common/pure/PendingOrderNotification/PendingOrderNotification.test.ts:43]: Would be great to use given/when/then format for test cases

File: apps/cowswap-frontend/src/common/pure/PendingOrderNotification/PendingOrderNotification.test.ts [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+const DAI_MAINNET = new Token(
+  ChainId.MAINNET,
+  '0x6b175474e89094c44da98b954eedeac495271d0f',
+  18,
+  'DAI',
+  'Dai Stablecoin'
+)
+
+const wethAmount = CurrencyAmount.fromRawAmount(WETH_MAINNET, 1e18)
+const daiAmount = CurrencyAmount.fromRawAmount(DAI_MAINNET, 2000 * 1e18)
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/common/pure/PendingOrderNotification/PendingOrderNotification.test.ts:20]: Can you add to the amounts (or a different test if you prefer) a decimal amount, like 1.231 ETH or something

File: apps/cowswap-frontend/src/common/pure/PendingOrderNotification/index.tsx [hunk 12]
```diff
@@ -28,26 +28,60 @@ const ORDER_TYPE_TITLES: Record<UiOrderType, string> = {
...
+
+  const baseMessage = (() => {
+    const isSellOrder = kind === OrderKind.SELL
+
+    if (isSellOrder) {
+      return `Sell ${inputAmountElement} for at least ${outputAmountElement}`
+    } else {
+      return `Buy ${outputAmountElement} for at most ${inputAmountElement}`
+    }
+  })()
```
Comment (alfetopito) [inline apps/cowswap-frontend/src/common/pure/PendingOrderNotification/index.tsx:70]: Nitpick: do a helper fn instead an inline one

**Comment by alfetopito** [comment]
I'm a bit concerned on how much prop drilling there is to pass `isSafeWallet` all the way down to the event emitter.
Is this a symptom of the code mess we have or could this somehow be done without all of this?
Is this a thing that we should have in a global state for being more integral to the core of the app?

File: apps/cowswap-frontend/src/common/pure/PendingOrderNotification/PendingOrderNotification.test.ts (hunk 10 above)
Comment (anxolin) [reply apps/cowswap-frontend/src/common/pure/PendingOrderNotification/PendingOrderNotification.test.ts:43]: True, I will change

**Comment by anxolin** [comment]
> I'm a bit concerned on how much prop drilling there is to pass isSafeWallet all the way down to the event emitter.
Is this a symptom of the code mess we have or could this somehow be done without all of this?
Is this a thing that we should have in a global state for being more integral to the core of the app?

Agree, I'm too. There's definitely an architecture problem. I expect the next events to be simpler. 
We have a global state, we just can't access from the middleware. 

In this case, I feel I had to pass the `isSafeWallet` to avoid racing conditions
- you get the flag from a global state, and it might have changed by the time you process the event

File: apps/cowswap-frontend/src/common/pure/PendingOrderNotification/PendingOrderNotification.test.ts (hunk 11 above)
Comment (anxolin) [reply apps/cowswap-frontend/src/common/pure/PendingOrderNotification/PendingOrderNotification.test.ts]: image

done a6026f029c1cf5795ea0ff55d58fb5804ee0e1ee

File: apps/cowswap-frontend/src/common/pure/PendingOrderNotification/index.tsx (hunk 12 above)
Comment (anxolin) [reply apps/cowswap-frontend/src/common/pure/PendingOrderNotification/index.tsx:70]: I'm happy to do if you prefer, but I did it IIFE because the function is simple enough, and this way the logic on how we compose the message is closer together. Its more analogous to the code in the component version of it

I know you don't love this pattern, just confirm me you feel I should do it, and I will move it. Otherwise I slightly prefer to have it self contained and specially this part is the gist of the function, so i prefer to not move it away

**Review by alfetopito — APPROVED** [review]
